python/Ganga/GPIDev/Credentials2/ICredentialInfo.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import absolute_import
+
+import copy
+import os
+from abc import ABCMeta, abstractmethod
+from datetime import datetime, timedelta
+from functools import wraps
+
+import Ganga.Utility.logging
+
+from .exceptions import CredentialsError
+
+logger = Ganga.Utility.logging.getLogger()
+
+
+def cache(method):
+    """
+    The cache decorator can be applied to any method in an ``ICredentialInfo` subclass.
+    It stores the return value of the function in the ``self.cache`` dictionary
+    with the key being the name of the function.
+    The cache is invalidated if ``os.path.getmtime(self.location)`` returns
+    greater than ``self.cache['mtime']``, i.e. if the file has changed on disk
+
+    Not having to call the external commands comes at the cost of calling ``stat()``
+    every time a parameter is accessed but this is still a saving of many orders of
+    magnitude.
+    """
+    @wraps(method)
+    def wrapped_function(self, *args, **kwargs):
+
+        # If the mtime has been changed, clear the cache
+        if os.path.exists(self.location):
+            mtime = os.path.getmtime(self.location)
+            if mtime > self.cache['mtime']:
+                self.cache = {'mtime': mtime}
+        else:
+            self.cache = {'mtime': 0}
+
+        # If entry is missing from cache, repopulate it
+        # This will run if the cache was just cleared
+        if method.func_name not in self.cache:
+            self.cache[method.func_name] = method(self, *args, **kwargs)
+
+        return self.cache[method.func_name]
+    return wrapped_function
+
+
+class ICredentialInfo(object):
+    """
+    The interface for all credential types.
+    Each object covers one credential file exactly.
```

Comment:
We don't have any credentials that need this, but is there any way at all to get this to work with a Credential that needs multiple files? Just interested to see if this cannot work at all or could be made to work somehow - just in case 😄